### DISCUSSION TITLE
Looking for guidance on how to navigate/combine Encounters and Vitals

### DISCUSSION THREAD

**Discussion by johnkwaters** [FirelyTeam/firely-net-sdk#1972] [Q&A] [answered]
I am trying to fetch Encounters for a patient with their associated vitals, and not quite getting the model or pattern, looking for some guidance on where to read up on this.

First I tried this, and also Include instead of ReverseInclude, but don't seem to get any Observations inside the Encounters.

    var esp = new SearchParams("subject", patientId)
        .ReverseInclude("Observations");
    var encounters = await client.SearchAsync<Encounter>(esp);

Then I tried this more convoluted approach

    var sp = new SearchParams("subject", patientId)
        .Include("Observation.encounter");
    var observations = await client.SearchAsync<Observation>(sp);
    var obsList = observations.Entry.ByResourceType<Observation>();

    var vitals = (
        from o in obsList
        let code = o.Category?.FirstOrDefault()?.Coding?.FirstOrDefault()?.Code
        where code is "vital-signs" 
        select o ).ToList();

Was then going to iterate over each vital sign Obeservation and match it to the Encounter. I do get the Encounter object populated in the Observation, but it has no Identifier, so then I have a fuzzy match on Encounter.URL which does contain the Encounter ID. And this LINQ query to weed out the Observations that are vital signs seems really unwieldy!

Help me... I can tell I am not approaching this right, but don't know where to find this kind of information in the docs...

Thanks!

**Comment by mbaltus** [comment, marked as answer]
Why not change your search to ask only for vital signs Observations? The category code that you try to filter on is also available as search parameter:

``` csharp
var sp = new SearchParams("subject", patientId);
sp.Add("category", "vital-signs");
```

So after your two searches you will have Encounters for a Patient and you will have a list of vital signs for that Patient. Am I correct in understanding that the next steps are to determine during which Encounter a vital sign was observed/recorded, so you can set the Observation.encounter to reference the Encounter?
If so, you should be able to use the Encounter's identity to set the reference, like this:

``` csharp
obs.Encounter = new ResourceReference (enc.ResourceIdentity().WithoutVersion().ToString());
```

Not quite sure what you mean by "but it has no Identifier, so then I have a fuzzy match on Encounter.URL which does contain the Encounter ID", so maybe I've skipped over an important detail here. Please clarify further, if my approach does not work.

**Reply by johnkwaters** [reply]
Thanks, I tried the search suggested above and that does cut out one step. For part two the problem is that when I add 

.Include("Observation.encounter");

To the Observation search, I DO get the Encounter object for each Observation - BUT the Encounter.Identifier is null in that object.

**Reply by johnkwaters** [reply]
![image](https://user-images.githubusercontent.com/3422637/154301457-db354055-dbc6-4e67-9d50-6850ce1cf7c5.png)

**Reply by johnkwaters** [reply]
Above is a vital, with it's Encounter object, and a null Identifier on the Encounter (and the URL that contains the encounter identifier)

**Reply by johnkwaters** [reply]
also - rather than fetch first the encounters and then the vitals and then try to correlate the two - is there not a way to get the vitals downloaded with each encounter - i.e. the children, in one call?

**Reply by johnkwaters** [reply]
also the Observation.Encounter seems to be a ResourceReference, not a Resource, so 

![image](https://user-images.githubusercontent.com/3422637/154302321-22f16995-7dd2-44dd-939e-2c6cdfcfb733.png)

**Reply by johnkwaters** [reply]
And I see REST API examples like this:
GET [base]/Patient/1234/?_type=Observation,Encounter

Is there a corresponding .NET syntax to make one call to get multiple resources for a given subject, rather than doing multiple calls? Which I am now doing in parallel with Task.WhenAll, for performance, but a 'bulk download' would be preferable

**Reply by mbaltus** [reply]
> Above is a vital, with it's Encounter object, and a null Identifier on the Encounter (and the URL that contains the encounter identifier)

You're confusing identifiers with technical ids. The technical identity of a resource takes the form of a URL. The technical id will be in the Resource.Id field. If the resource has any business/administrative identifiers, those would end up in the Resource.Identifier field.
So, the URL you mention does _not_ contain the identifier, but it contains the id, so it is very well possible that the identifier field is empty.

> also - rather than fetch first the encounters and then the vitals and then try to correlate the two - is there not a way to get the vitals downloaded with each encounter - i.e. the children, in one call?

Yes, you do that by using the Include on the search, but this only works if your Observation has a link to an Encounter - in your other ticket you mentioned that it did not give you the Encounters. That would indicate that they are not correlated in the FHIR data you retrieve, and in turn that means you cannot use the correlation to retrieve all in one call.
Note that for (Reverse)Include you use a ':' not a '.' to separate the resource type from the search parameter:

``` csharp
var sp = new SearchParams("subject", patientId);
sp.Add("category", "vital-signs");
sp.Include("Observation:encounter") // or, if there is a link from Encounter to Observation: sp.ReverseInclude("Encounter:reason-reference");
```

> also the Observation.Encounter seems to be a ResourceReference, not a Resource

Correct. FHIR does not embed resources inline in another resource, but works with references instead. The objects are considered to be standalone objects which you can retrieve separately from each other, or, if you have the references correctly filled, you can do a search like just discussed.

**Reply by johnkwaters** [reply]
Thanks. But the Vital (Observation) _does_ have an Encounter object... I just can't find out how to correlate that Object to the Encounters in my list of Encounters.

**Reply by johnkwaters** [reply]
![image](https://user-images.githubusercontent.com/3422637/154309491-adaf50da-2c4b-412c-938b-6489942596dc.png)

**Reply by johnkwaters** [reply]
Is this what I am looking for?

![image](https://user-images.githubusercontent.com/3422637/154309840-81a2e4ca-c042-4dfd-8652-32d224bd3d2d.png)

**Reply by johnkwaters** [reply]
Like this?

Is there a way, a helper, to use to compare that "Encounter/{someid}" with the {id} as shown below? Or do I remove the Encounter/ and compare the rest of the string to match them?

![image](https://user-images.githubusercontent.com/3422637/154312385-c84fc29c-7571-4c6b-be1e-9ee128916863.png)

**Reply by mbaltus** [reply]
Yes, that Reference value should match one of the Encounters in your Encounter bundle. If you only have a relative reference, you can use ResourceIdentity to create the absolute reference so you can compare that with the fullUrl fields in your Encounter Bundle:

``` csharp
var encUrl = new ResourceIdentity(obs.Encounter.Reference).WithBase(c.Endpoint);
var result = encounters.findEntry(encUrl);
```

Since the obs.Encounter is filled in, the search in my previous example should work to do just one call.

**Reply by mbaltus** [reply]
I used c.Endpoint to get my FhirClient's base url, but you could use a hardcoded string as well.

**Reply by johnkwaters** [reply]
Thanks! So I tried fetching multiple things related to a patient in one call with Includes, like this,

    var patients = await client.SearchAsync<Patient>(
        new SearchParams("_id", patientId)
            .Include("Condition")
            .Include("Observation")
            .Include("AllergyIntolerance"));
    foreach (var patient1 in patients.Entry.ByResourceType<Patient>())
    {
        
    }
    foreach (var patient1 in patients.Entry.ByResourceType<Observation>())
    {

    }
    foreach (var patient1 in patients.Entry.ByResourceType<Condition>())
    {

    }
    foreach (var patient1 in patients.Entry.ByResourceType<AllergyIntolerance>())
    {

    }

But all those collections of other things are empty. What am I doing wrong?

**Reply by mbaltus** [reply]
> And I see REST API examples like this: GET [base]/Patient/1234/?_type=Observation,Encounter
> 
> Is there a corresponding .NET syntax to make one call to get multiple resources for a given subject, rather than doing multiple calls? Which I am now doing in parallel with Task.WhenAll, for performance, but a 'bulk download' would be preferable

I don't think the SDK supports compartment searches at this time, but you could try a `WholeSystemSearch` and supply the `_type` and `patient` parameters.  Or, if the server supports the $everything operation, you can use `FetchPatientRecord` for getting all resources linked to a patient.

**Reply by johnkwaters** [reply]
I think it does support $everything - I can try that. FetchPatientRecord takes a Uri - how do I make one of those from a patient Id?

**Reply by mbaltus** [reply]
In your last example, your includes are incomplete. You should always list the resource type _and_ the search parameter. Also, they should be reverse includes, because the references are **from** the other type **to** the Patient, so your search goes against the direction. See http://hl7.org/fhir/search.html#revinclude for more info on the syntax.

This should work:
```
   .ReverseInclude("Condition:patient")
```

Edited to add: the reason for also listing the search parameter is that the Patient can be linked in other fields than just the subject field. For example you could want to include Observations where the Patient is the performer but not necessarily the subject: `.ReverseInclude("Observation:performer")`

**Reply by mbaltus** [reply]
> I think it does support $everything - I can try that. FetchPatientRecord takes a Uri - how do I make one of those from a patient Id?

Use the ResourceIdentity for that:

```
c.FetchPatientRecord(ResourceIdentity.Build("Patient", patientId));
```

**Reply by johnkwaters** [reply]
I guess everything doesn't work :-)

![image](https://user-images.githubusercontent.com/3422637/154317023-a2ba2418-8071-41c4-9b7c-40cf9d02ace2.png)

**Reply by johnkwaters** [reply]
So those reverse includes DO return the other related elements! Now... how do I for those observations included limit them to be vital signs? 

    var patients = await client.SearchAsync<Patient>(
        new SearchParams("_id", patientId)
            .ReverseInclude("Condition:patient")
            .ReverseInclude("Observation:patient")
            .ReverseInclude("AllergyIntolerance:patient"));

**Reply by mbaltus** [reply]
Ah, sorry, that is where you run into a limit on what FHIR search can do for you... you cannot filter the included resources in the search call. You can only filter on the resource type that is the main search type.

You could form your search the other way around, and use iteration on the includes to get all the data you want to add:
``` csharp
var observations = await client.SearchAsync<Observation>(
    new SearchParams("patient", patientId)
        .Add("category", "vital-signs")
        .Include("Observation:patient")
        .ReverseInclude("Condition:patient", IncludeModifier.Iterate)
        .ReverseInclude("AllergyIntolerance:patient", IncludeModifier.Iterate));
```
Edited to add the Patient resource as well, since that is needed for the iterations.

**Reply by johnkwaters** [reply]
But then if the patient had no vitals I wouldn't get anything right? I might go for the utility of getting everything in one call, and then narrow down the vitals locally. My other approach was doing all the calls in parallel, which was fairly fast. Any recommendations on best practice to get all the data for a patient as quickly as possible (since everything doesn't work). Use the includes, the parallel searches, or something else?

    var patientTask = client.ReadAsync<Patient>($"Patient/{patientId}");
    var esp = new SearchParams("subject", patientId)
        .ReverseInclude("Observations");
    var encountersTask = client.SearchAsync<Encounter>(esp);
    var immunizationsTask = client.SearchAsync<Immunization>(new SearchParams("patient", patientId));
    var conditionsTask  = client.SearchAsync<Condition>(new SearchParams("subject", patientId));
    var allergiesTask = client.SearchAsync<AllergyIntolerance>(new SearchParams("patient", patientId));
    var sp = new SearchParams("subject", patientId)
        .Where($"subject={patientId}")
        .Where("category=vital-signs")
        .Include("Observation.encounter");
    var observationsTask = client.SearchAsync<Observation>(sp);
    var proceduresTask = client.SearchAsync<Procedure>(new SearchParams("patient", patientId));

    await System.Threading.Tasks.Task.WhenAll(patientTask, encountersTask, immunizationsTask, conditionsTask, allergiesTask,
        observationsTask, proceduresTask);

**Reply by mbaltus** [reply]
> But then if the patient had no vitals I wouldn't get anything right?

Correct, so you'll have to think about what you would need, and maybe do some extra client side filtering like you suggest.

I don't really have recommendations on which approach is best. Includes could be slower since this could be one big query with joins on the database, and if that's not up to it, you'll see performance drop. On the other hand, a lot of calls in parallel could lead to more network traffic, and also lots of parallel queries on the database. Maybe the FHIR community (https://chat.fhir.org) has more ideas on this. Or you could do some tests for both and time them.

You didn't ask, but I would recommend to change that esp SearchParams, since there is still an incorrect include on that. And, for each resource that can have a Patient as subject, there is a `patient` parameter. So all searches can be done with `new SearchParams("patient", patientId)`. For most of your searches that means you can re-use one SearchParams if you want to make your code look a bit cleaner.

**Reply by johnkwaters** [reply]
I did a quick test, and the big search for 11 different resources took 26s, while the parallel version took 23, so they seem quite comparable.

**Reply by johnkwaters** [reply]
Yeah, end result with these changes is 25s vs 23s, program is:

`
#define xPARALLELL

using System.Diagnostics;
using System.Net.Http.Headers;
using System.Text;
using FHIR2HC;
using FHIR2HC.HC;
using Hl7.Fhir.Model;
using Hl7.Fhir.Rest;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;
using Immunization = Hl7.Fhir.Model.Immunization;

// TODO - add rest of elements
// Call HC API to look up existing record

var builder = new ConfigurationBuilder()
    .AddJsonFile($"appsettings.json", true, true);
var config = builder.Build();

var FHIRUrl = config["fhir.url"];
var username = config["fhir.username"];
var password = config["fhir.password"];

var hcUrl = config["hc.url"];
var hcUserName = config["hc.username"];
var hcPassword = config["hc.password"];
var hcMatrix = config["hc.matrix"];
var hcIncident = config["hc.incident"];
var hcAgency = config["hc.agency"];
var enablesend = bool.Parse(config["hc.enablesend"]);

var client = new FhirClient(
    FHIRUrl, 
    new FhirClientSettings
    {
        Timeout = 10000,
        PreferredFormat = ResourceFormat.Json,
        VerifyFhirVersion = false,
        PreferredReturn = Prefer.ReturnMinimal
    }, 
    new AuthorizationMessageHandler(new AuthenticationHeaderValue("Basic", Convert
            .ToBase64String(Encoding.GetEncoding("ISO-8859-1")
                .GetBytes($"{username}:{password}"))))
    );

HttpClient hcClient = null;
if (enablesend)
{
    var handler = new AuthorizationMessageHandler(new AuthenticationHeaderValue("Basic", Convert
        .ToBase64String(Encoding.GetEncoding("ISO-8859-1")
            .GetBytes($"{hcUserName}:{hcPassword}"))));
    hcClient = new HttpClient(handler)
    {
        BaseAddress = new Uri(hcUrl)
    };
    hcClient.DefaultRequestHeaders.Add("TimeZone_StandardName", "America/Los_Angeles");
    hcClient.DefaultRequestHeaders.Accept.Clear();
    hcClient.DefaultRequestHeaders.Accept.Add(
        new MediaTypeWithQualityHeaderValue("application/json"));


    Console.WriteLine("Logging in to HC...");
    var tokenResponse = await hcClient.GetAsync("1.0/token");
    tokenResponse.EnsureSuccessStatusCode();
    var token = await tokenResponse.Content.ReadAsAsync<TokenResponse>();
    handler.Authorization = new AuthenticationHeaderValue("Session", token.access_token);
    Console.WriteLine("Logged in to HC...");
}

var patientIds = await File.ReadAllLinesAsync("PatientIds.txt");
Console.WriteLine($"Found {patientIds.Length} records to fetch");

var saveParams = new SaveParams
{
    MatrixName = hcMatrix,
    EnsureChangesStayInFilter = true,
    InsertedEntities = new List<InsertedEntity>()
};

var counter = 0;

var sw = new Stopwatch();
sw.Start();

foreach (var patientId in patientIds)
{
    var p = new InsertedEntity()
    {
        PatientID = patientId,
        Incident_Name = hcIncident,
        PT_Agency_Name = hcAgency,
        ID = -(++counter)
    };
    saveParams.InsertedEntities.Add(p);

    Console.WriteLine($"Fetching record {patientId}");

    var psp = new SearchParams("patient", patientId);
    var osp = psp
        .Where("category=vital-signs")
        .Include("Observation.encounter");
    var observationsTask = client.SearchAsync<Observation>(osp);

#if PARALLELL    
    var patientTask = client.ReadAsync<Patient>($"Patient/{patientId}");
    var encountersTask = client.SearchAsync<Encounter>(psp);
    var immunizationsTask = client.SearchAsync<Immunization>(psp);
    var conditionsTask  = client.SearchAsync<Condition>(psp);
    var allergiesTask = client.SearchAsync<AllergyIntolerance>(psp);
    var proceduresTask = client.SearchAsync<Procedure>(psp);
    await System.Threading.Tasks.Task.WhenAll(
        patientTask, 
        encountersTask,immunizationsTask,conditionsTask, allergiesTask,
        observationsTask,
        proceduresTask
    );

#else
    var patientsTask = client.SearchAsync<Patient>(
        new SearchParams("_id", patientId)
            .ReverseInclude("Encounter:patient")
            .ReverseInclude("Condition:patient")
            .ReverseInclude("Immunization:patient")
            .ReverseInclude("Procedure:patient")
            .ReverseInclude("AllergyIntolerance:patient"));
    await System.Threading.Tasks.Task.WhenAll(
        patientsTask,
        observationsTask
        );

#endif

    // $everything throws a server error
    //var p2 = await client.FetchPatientRecordAsync(
    //    ResourceIdentity.Build("Patient", patientId));

#if PARALLELL
    var patient = patientTask.Result;
#else
    var patients = patientsTask.Result;
    var patient = patients.Entry.ByResourceType<Patient>().SingleOrDefault();
#endif

    var humanName = patient.Name.First();
    var givenNames = string.Join(" ", humanName.Given);
    Console.WriteLine($"Found {givenNames} {humanName.Family}");

    p.FirstName = givenNames;
    p.LastName = humanName.Family;
    p.DOB = patient.BirthDateElement.ToDateTimeOffset().Value.Date;
    p.Sex_Name = patient.Gender switch {
        AdministrativeGender.Female => "Female",
        AdministrativeGender.Male => "Male",
        _ => null
    };

#if PARALLELL
    var encounters = encountersTask.Result.Entry.ByResourceType<Encounter>().ToList();
#else
    var encounters = patients.Entry.ByResourceType<Encounter>().ToList();
#endif
    Console.WriteLine($" {encounters.Count} Encounters:");
    foreach (var e in encounters)
    {
        var description = string.Join(", ", e.Type.Select( c => c.Text));
        Console.WriteLine($"   {description}, status {e.Status}, id: {e.Id}, from {e.Period.Start} to {e.Period.End}");
    }

#if PARALLELL
    var immunizations = immunizationsTask.Result.Entry.ByResourceType<Immunization>().ToList();
#else
    var immunizations = patients.Entry.ByResourceType<Immunization>().ToList();
#endif

    Console.WriteLine($" {immunizations.Count} Immunizations:");
    foreach (var i in immunizations)
    {
        Console.WriteLine($"    {i.ExpirationDate} {i.VaccineCode.Text}");
    }

#if PARALLELL
    var conditions = conditionsTask.Result.Entry.ByResourceType<Condition>().ToList();
#else
    var conditions = patients.Entry.ByResourceType<Condition>().ToList();
#endif
    Console.WriteLine($" {conditions.Count} Conditions):");
    foreach (var c in conditions)
    {
        var code = c.Code.Coding.First();
        Console.WriteLine(
            $"   {c.Code.Text} onset {c.Onset}, severity {c.Severity?.Text} ({code.Display} {code.Code})");
    }

#if PARALLELL
    var allergies = allergiesTask.Result.Entry.ByResourceType<AllergyIntolerance>().ToList();
#else
    var allergies = patients.Entry.ByResourceType<AllergyIntolerance>().ToList();
#endif
    
    Console.WriteLine($" {allergies.Count} Allergies:");
    foreach (var a in allergies)
    {
        var categories = string.Join(" ", a.Category);
        var code = a.Code.Coding.First();
        Console.WriteLine($"   {a.Code.Text} recorded {a.RecordedDate}, categories: {categories}, ({code.Display} {code.Code})");
        p.Allergies.Add(
            new Allergy { PTS2_ALG_Allergy_Name = a.Code.Text });
    }

    var observations = observationsTask.Result;
    Console.WriteLine(" Observations:");

    foreach (var v in observations.Entry.ByResourceType<Observation>())
    {
        var codings = string.Join(", ", v.Code.Coding.Select(c => $"{c.Display} {c.Code}"));
        if (v.Value == null)
        {
            continue;
        }
        var val = v.Value switch
        {
            Quantity q => $"{q.Value} {q.Unit}",
            _ => v.Value.ToString()
        };
        Console.WriteLine($"     {codings}, effective {v.Effective} {val}, status {v.Status}, encounter {v.Encounter?.Reference}");
        if (v.Code.Text == "Body weight")
        {
            p.Weight = (int)(v.Value as Quantity).Value;
        }
    }

#if PARALLELL
    var procedures = proceduresTask.Result.Entry.ByResourceType<Procedure>().ToList();
#else
    var procedures = patients.Entry.ByResourceType<Procedure>().ToList();
#endif
    
    Console.WriteLine($" {procedures.Count} Procedures:");
    foreach (var proc in procedures)
    {
        var categories = string.Join(" ", proc.Category);
        var code = proc.Code.Coding.First();
        var period = proc.Performed as Period;
        Console.WriteLine($"   {proc.Code.Text} encounter {proc.Encounter.Display}, performed {period.Start} {period.End}, status {proc.Status}");
    }

    Console.WriteLine("-----------------------------------");

    var msg = JsonConvert.SerializeObject(saveParams);

    if (enablesend)
    {
        var saveResponse = await hcClient.PostAsJsonAsync("dynamic/1.0/save", saveParams);
        saveResponse.EnsureSuccessStatusCode();

        Console.WriteLine("Logging out from HC...");
        await hcClient.GetAsync("sessions/1.0/logout");
    }


}
sw.Stop();
Console.WriteLine($"Elapsed time {sw.Elapsed.Seconds}s");

Console.ReadKey();
`

**Reply by mbaltus** [reply]
Looks promising! I'll close this issue, since your questions seem to be answered, but please feel free to reopen or create a new one.
